<a href="https://colab.research.google.com/github/gulshan-katankar/language_modelling/blob/main/model_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import torch
import torch.nn.functional as F
import matplotlib.pyplot as plt
%matplotlib inline

In [2]:
import requests

url = "https://raw.githubusercontent.com/gulshan-katankar/language_modelling/refs/heads/main/names.txt"

words = requests.get(url).text.splitlines()

In [3]:
words[:8]

['emma', 'olivia', 'ava', 'isabella', 'sophia', 'charlotte', 'mia', 'amelia']

In [4]:
len(words)

32033

In [5]:
#build the vocab of characters and mappings

chars = sorted(list(set(''.join(words))))
stoi = {s:i+1 for i,s in enumerate(chars)}
stoi['.'] = 0
itos = {i:s for s,i in stoi.items()}
print(itos)

{1: 'a', 2: 'b', 3: 'c', 4: 'd', 5: 'e', 6: 'f', 7: 'g', 8: 'h', 9: 'i', 10: 'j', 11: 'k', 12: 'l', 13: 'm', 14: 'n', 15: 'o', 16: 'p', 17: 'q', 18: 'r', 19: 's', 20: 't', 21: 'u', 22: 'v', 23: 'w', 24: 'x', 25: 'y', 26: 'z', 0: '.'}


In [7]:
# build the dataset

block_size = 3 # context length: how many characters do we tako to predict the next one?
x, y = [], []
for w in words [:5] :

  print(w)
  context = [0] * block_size #start with the padded context of 0
  for ch in w + '.': #iterate over all the char
    ix = stoi[ch]
    x. append(context) #stores the context
    y.append(ix) #stores the labels for the context
    print(''.join(itos[i] for i in context), ' -- >', itos[ix])
    context = context [1:] + [ix] # crop and append

x = torch. tensor(X)
y = torch. tensor(Y)

emma
...  -- > e
..e  -- > m
.em  -- > m
emm  -- > a
mma  -- > .
olivia
...  -- > o
..o  -- > l
.ol  -- > i
oli  -- > v
liv  -- > i
ivi  -- > a
via  -- > .
ava
...  -- > a
..a  -- > v
.av  -- > a
ava  -- > .
isabella
...  -- > i
..i  -- > s
.is  -- > a
isa  -- > b
sab  -- > e
abe  -- > l
bel  -- > l
ell  -- > a
lla  -- > .
sophia
...  -- > s
..s  -- > o
.so  -- > p
sop  -- > h
oph  -- > i
phi  -- > a
hia  -- > .


/tmp/ipykernel_795/1879530691.py:16: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.detach().clone() or sourceTensor.detach().clone().requires_grad_(True), rather than torch.tensor(sourceTensor).
  x = torch. tensor(X)
/tmp/ipykernel_795/1879530691.py:17: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.detach().clone() or sourceTensor.detach().clone().requires_grad_(True), rather than torch.tensor(sourceTensor).
  y = torch. tensor(Y)


In [8]:
x.shape, x.dtype, y.shape, y.dtype

(torch.Size([32, 3]), torch.int64, torch.Size([32]), torch.int64)

from 5 words we have a dataset of 32 examples, where each neural net have an input of size 3 they are integers

the labels too are integers

In [9]:
x

tensor([[ 0,  0,  0],
        [ 0,  0,  5],
        [ 0,  5, 13],
        [ 5, 13, 13],
        [13, 13,  1],
        [ 0,  0,  0],
        [ 0,  0, 15],
        [ 0, 15, 12],
        [15, 12,  9],
        [12,  9, 22],
        [ 9, 22,  9],
        [22,  9,  1],
        [ 0,  0,  0],
        [ 0,  0,  1],
        [ 0,  1, 22],
        [ 1, 22,  1],
        [ 0,  0,  0],
        [ 0,  0,  9],
        [ 0,  9, 19],
        [ 9, 19,  1],
        [19,  1,  2],
        [ 1,  2,  5],
        [ 2,  5, 12],
        [ 5, 12, 12],
        [12, 12,  1],
        [ 0,  0,  0],
        [ 0,  0, 19],
        [ 0, 19, 15],
        [19, 15, 16],
        [15, 16,  8],
        [16,  8,  9],
        [ 8,  9,  1]])

In [10]:
y

tensor([ 5, 13, 13,  1,  0, 15, 12,  9, 22,  9,  1,  0,  1, 22,  1,  0,  9, 19,
         1,  2,  5, 12, 12,  1,  0, 19, 15, 16,  8,  9,  1,  0])

In [11]:
c = torch.randn((27,2)) #this is the embedding table, every char will have a 2 dim embedding

In [12]:
c #this is randomly generated as a model param, and will be optimized further

tensor([[ 1.2832e+00, -5.5841e-01],
        [-1.7665e-01,  1.3457e+00],
        [ 1.2621e+00,  2.9356e-01],
        [-1.8683e-01, -3.9807e-01],
        [ 1.8270e-01, -4.9373e-01],
        [-1.2772e+00,  1.3232e+00],
        [-4.7474e-02, -2.0602e-01],
        [ 5.3236e-02,  1.4040e+00],
        [ 4.1752e-01, -9.7645e-02],
        [ 1.3166e+00, -2.6769e-01],
        [-2.1983e-03,  2.2881e-01],
        [ 1.3150e+00, -5.3842e-01],
        [ 2.5157e-02,  2.6276e+00],
        [-3.0813e-01,  1.8799e+00],
        [ 2.9645e-02, -5.1233e-01],
        [ 2.4169e+00, -3.9956e-01],
        [-1.1479e+00, -2.0835e-01],
        [ 9.6115e-01,  1.2409e+00],
        [ 8.6723e-02, -6.6782e-01],
        [-2.9608e-01,  9.3969e-01],
        [ 1.9885e-01,  4.4112e-01],
        [-4.8309e-01, -1.1641e+00],
        [-7.5322e-01,  5.6410e-01],
        [ 1.3311e+00,  7.4322e-01],
        [-2.8638e-02, -8.0578e-01],
        [ 3.1049e-01,  4.9216e-01],
        [ 4.5438e-01,  3.2429e+00]])

so now we willgenerate embeddings for the x using the random embedding table we generated above

In [16]:
emb = c[x]
emb

tensor([[[ 1.2832, -0.5584],
         [ 1.2832, -0.5584],
         [ 1.2832, -0.5584]],

        [[ 1.2832, -0.5584],
         [ 1.2832, -0.5584],
         [-1.2772,  1.3232]],

        [[ 1.2832, -0.5584],
         [-1.2772,  1.3232],
         [-0.3081,  1.8799]],

        [[-1.2772,  1.3232],
         [-0.3081,  1.8799],
         [-0.3081,  1.8799]],

        [[-0.3081,  1.8799],
         [-0.3081,  1.8799],
         [-0.1766,  1.3457]],

        [[ 1.2832, -0.5584],
         [ 1.2832, -0.5584],
         [ 1.2832, -0.5584]],

        [[ 1.2832, -0.5584],
         [ 1.2832, -0.5584],
         [ 2.4169, -0.3996]],

        [[ 1.2832, -0.5584],
         [ 2.4169, -0.3996],
         [ 0.0252,  2.6276]],

        [[ 2.4169, -0.3996],
         [ 0.0252,  2.6276],
         [ 1.3166, -0.2677]],

        [[ 0.0252,  2.6276],
         [ 1.3166, -0.2677],
         [-0.7532,  0.5641]],

        [[ 1.3166, -0.2677],
         [-0.7532,  0.5641],
         [ 1.3166, -0.2677]],

        [[-0.7532,  0

if you read the above embeddings carefully you can see pytorch assigned the same value the index have from the embedding table as well as the the mapping we did

for eg :

. . e -> [[ 1.2832, -0.5584],
         [ 1.2832, -0.5584],
         [-1.2772,  1.3232]],

here . has a 0 index in x and e has 5th index so the embeddings we got also have the same index value from the embeddings table c

In [17]:
emb.shape

torch.Size([32, 3, 2])

now we have the embeddings for all of the 32, 3 integers

In [21]:
w1 = torch.randn((6, 100)) #6 coz we have 3 input size and 2 dim, and 100 are the no of neurons
b1 = torch.randn(100)

**hidden layer calculation**

In [22]:
h = torch.tanh(emb.view(-1,6) @ w1 + b1) # -1 coz python will automatically interpret what should be the size here

In [23]:
h

tensor([[-0.9930, -0.7677,  0.9996,  ...,  0.7064, -0.9998,  0.9971],
        [ 0.9900, -0.9987,  1.0000,  ...,  0.9995,  0.9756,  0.9889],
        [ 0.9999, -0.9043,  0.9931,  ...,  1.0000,  0.7635,  0.9991],
        ...,
        [ 0.9907, -0.9940,  0.8267,  ...,  0.9999, -0.8935,  0.9998],
        [-0.9994,  0.9895,  0.4982,  ..., -0.8570, -0.9999, -0.2365],
        [ 0.1723, -0.8081,  1.0000,  ...,  0.9681, -0.7498,  0.9800]])

In [24]:
h.shape

torch.Size([32, 100])

In [26]:
w2 = torch.randn((100, 27)) #the input now is 100 and the output would be 27 coz we have 27 char as the next possible output
b2 = torch.randn(27)

In [27]:
logits = h @ w2 + b2 # the output of the neural nets

In [28]:
logits

tensor([[-9.6241e-01,  8.7117e+00, -4.1533e+00,  6.3783e+00,  4.1085e+00,
          3.4683e+00, -3.3922e+00,  3.4957e+00,  8.0812e+00, -5.6316e+00,
         -3.2044e+00,  5.8886e+00, -3.8442e+00,  1.5941e+01,  3.2525e+00,
          1.0316e+01,  3.2573e+00, -8.3507e+00, -4.9445e-01,  7.5676e+00,
          1.1000e+01, -1.2190e+00,  4.9274e+00, -3.6955e+00, -8.9107e+00,
          1.4856e+01, -3.0137e+00],
        [-4.6257e+00, -1.2723e+00, -1.4374e+00,  6.3743e+00,  2.3242e+00,
          5.3842e+00,  1.7160e+00,  6.4711e+00, -2.4057e+00, -1.1855e+00,
          1.0582e+01, -1.5583e+00, -1.5373e+01, -4.2509e+00,  4.1009e+00,
         -1.4523e+01, -2.6783e+00, -5.0401e+00,  2.9115e+00,  1.8049e+01,
          2.9333e-01,  2.5004e+00,  8.3599e+00,  2.8896e-01, -1.3916e+01,
          1.1944e+01,  8.1105e+00],
        [-1.0134e+01,  6.0299e+00, -1.3943e+01, -7.2095e+00, -7.9961e-01,
          7.7055e-01,  2.9018e+00,  1.1434e+00, -3.7501e+00,  7.0194e+00,
          8.7619e+00, -1.1569e+01, -1.56

In [29]:
logits.shape

torch.Size([32, 27])

In [30]:
counts = logits.exp()

In [31]:
prob = counts / counts.sum(1, keepdims=True)

In [32]:
prob.shape

torch.Size([32, 27])

In [33]:
loss = -prob[torch.arange(32), y].log().mean()
loss

tensor(16.6342)

**clean version**

In [34]:
x.shape, y.shape

(torch.Size([32, 3]), torch.Size([32]))

In [35]:
g = torch.Generator().manual_seed(2147483647) # for reproducibility
C = torch.randn((27, 2), generator=g)
W1 = torch.randn((6, 100), generator=g)
b1 = torch.randn(100, generator=g)
W2 = torch.randn((100, 27), generator=g)
b2 = torch.randn(27, generator=g)
parameters = [C, W1, b1, W2, b2]

In [36]:
sum(p.nelement() for p in parameters) #no of params in total

3481

In [40]:
emb = C[x] # (32, 3, 2)
h = torch.tanh(emb.view(-1, 6) @ W1 +b1) # (32, 100)
logits = h @ W2 + b2 # (32, 27)
counts = logits.exp()
prob = counts / counts.sum(1, keepdims=True)
loss =- prob[torch.arange(32), y].log().mean()
loss

tensor(17.7697)